icu4c/source/i18n/zonemeta.cpp

```diff
@@ -389,6 +390,35 @@ ZoneMeta::getCanonicalCLDRID(const TimeZone& tz) {
     return getCanonicalCLDRID(tz.getID(tzID), status);
 }
 
+UnicodeString& U_EXPORT2
+ZoneMeta::getIanaID(const UnicodeString& tzid, UnicodeString& ianaID, UErrorCode& status) {
+    // First, get CLDR canonical ID
+    const char16_t *canonicalID = getCanonicalCLDRID(tzid, status);
+    if (U_FAILURE(status) || canonicalID == nullptr) {
+        ianaID.setToBogus();
+        return ianaID;
+    }
+    // Find IANA mapping if any.
+    UErrorCode tmpStatus = U_ZERO_ERROR;
+    UnicodeString tmpKey(canonicalID);
+    tmpKey.findAndReplace(UnicodeString("/"), UnicodeString(":"));
+    char keyBuf[ZID_KEY_MAX + 1];
+    /* int32_t keyLen = */ tmpKey.extract(0, tmpKey.length(), keyBuf, sizeof(keyBuf), US_INV);
+
+    StackUResourceBundle r;
+    ures_openDirectFillIn(r.getAlias(), nullptr, gKeyTypeData, &tmpStatus);
+    ures_getByKey(r.getAlias(), gIanaMapTag, r.getAlias(), &tmpStatus);
+    ures_getByKey(r.getAlias(), gTimezoneTag, r.getAlias(), &tmpStatus);
+    int32_t tmpLen = 0;
+    const char16_t* tmpIana = ures_getStringByKey(r.getAlias(), keyBuf, &tmpLen, &tmpStatus);
+    if (U_SUCCESS(tmpStatus)) {
+        ianaID.setTo(tmpIana, tmpLen);
+    } else {
+        ianaID.setTo(canonicalID, -1);
```

Comment:
Since the canonicalID is returned by the version which "overload method returns a persistent const char16_t*, which is guaranteed to persist    * (a pointer to a resource)."  of getCanonicalCLDRID() 

I think we should also use ianaID.setTo(true, canonicalID, -1);

https://github.com/unicode-org/icu/blob/cc2ddc0d113aa4db6efa6b071eca172e4a72f6cd/icu4c/source/i18n/zonemeta.h#L50

to save internal memory allocatioun for ianaID

Reply:
will make this change.

Reply:
done